### DISCUSSION TITLE
Concurrency/Multithreading Help

### DISCUSSION THREAD

**Discussion by AGogosha** [vlang/v#17230] [Questions and Answers] [answered]
At the recommendation of u/waozen under my original [Reddit post](https://www.reddit.com/r/vlang/comments/10u5b4j/concurrencymultithreading_help/) I am posting this here as well to hopefully get some help
This is gonna be a bit long so bear with me. To start I will give some context for my issue. I am quite new to V, I started learning V because I wanted to know something other than python and V has an extremely low barrier to entry. All you learn as a physics major is Python and I don't want to jump into Fortran for physics simulations. For an initial project, I decided to rewrite an old extra credit problem I did in one of my old physics classes. The problem was to program a simulation of a ball rolling around on a rotating table with different starting positions and then plot the path. The expected paths are shown in the following image.
![Python Output](https://user-images.githubusercontent.com/112112846/216859634-ac51126f-59ad-4bdb-b66b-eaf4626d94a7.png)

Now onto my issue. Note that all the math calculations are done in V then the data is outputted to a CSV file and then the data is plotted in python. The following paths were generated without multithreading and produced the correct output.
![V Correct Output](https://user-images.githubusercontent.com/112112846/216859649-ac8a85e0-0f34-4a15-b7f9-6b17d35a2789.png)

When I implement concurrency/multithreading as shown in the last example of [V concurrency documentation](https://github.com/vlang/v/blob/master/doc/docs.md#concurrency) I get the following output.
![V Incorrect Output](https://user-images.githubusercontent.com/112112846/216859670-6f154471-b464-4661-b8bc-ce893cada639.png)

Specifically, I implemented concurrency/multithreading it as follows:

```
mut threads := []thread []mv.Vec3[f64]{cap: init_vel.len}
    for i, vel in init_vel {
        threads << spawn cal_path(pos_init, vel, omega, delt_t, init_durations[i])
}
path := threads.wait()
```

Where cal_path calculates the path by calling several other functions in a loop and outputting the resulting position vectors into a single array. I don't know how this only breaks the last three simulations but it does and I'm not sure how to fix this. Any help would be greatly appreciated.

If you want to explore my not so great V code along with the CSV outputs, plots, python plotting code, and my original simulation code it can all be found [here](https://github.com/AGogosha/ball_simulation).

Another unrelated issue but still an interesting one is if I change all my variables to float 32's instead of float 64's. Everything breaks as you can see.
![f32 Weirdness Output](https://user-images.githubusercontent.com/112112846/216859706-ac62e83e-3ee3-40c6-a605-e2e6e1e599e0.png)

TLDR: Concurrency only breaks 3 out of the 6 simulations and I have no idea why. The code is [here](https://github.com/AGogosha/ball_simulation) if you want to read it.

**Comment by JalonSolov** [comment]
@ulises-jeremias @hungrybluedev

**Comment by hungrybluedev** [comment]
I faced a similar issue when I tried to multi-thread my RNG analysis code. There was some overlapping that led to the wrong values being updated. I think it was fixed for my use case.

Maybe a small reproducible example will help, like it did for my case.

**Comment by larpon** [comment, marked as answer]
Confirmed
![image](https://user-images.githubusercontent.com/768942/216918274-9e425aa3-10b9-487d-9e56-b3001a0aa270.png)

Since my initial thought was that it was a memory issue I tried commenting out the `cap: loop_length` [here](https://github.com/AGogosha/ball_simulation/blob/master/src/main.v#L48) which, on the next runs, gave better, though not perfect, results:
![image](https://user-images.githubusercontent.com/768942/216922463-7664e199-dcae-446a-ac98-b4e4e1454578.png)

So, at first glance, to me it looks like some overlapping memory issue when the array(s) are initialised in the thread.
This suspicion also gains weight since when running the code unmodified with `v -prealloc src/main.v` outputs the expected data:
![image](https://user-images.githubusercontent.com/768942/216924924-f51d4556-714a-48fd-865b-dd862af31912.png)

Maybe we need to look at memory allocation of arrays if they're inside threaded code.
